spharpy/spherical.py

```diff
@@ -1,94 +1,455 @@
+        - ``'sn3d'``: Uses the SN3D normalization (also known as Schmidt normalized).
+    """
+    def __init__(self, coords, basis_type='complex', channel_convention='acn', inv_transform_type=None,
+                 normalization='n3d'):
+        r"""
+        Initialize the SphericalHarmonics class.
+
+        Parameters
+        ----------
+        n_max : int
+            Maximum spherical harmonic order
+        coords : :doc:`pf.Coordinates <pyfar:classes/pyfar.coordinates>`
+            Coordinate object with sampling points for which the basis matrix is calculated
+        basis_type : str, optional
+            Type of spherical harmonic basis, either ``'complex'`` or ``'real'``. The default is ``'complex'``.
+        channel_convention : str, optional
+            Channel ordering convention, either ``'acn'`` or ``'fuma'``. The default is ``'acn'``.
+            (FuMa is only supported up to 3rd order)
+        inv_transform_type : str, optional
+            Inverse transform type, either ``'pseudo_inverse'`` or ``'quadrature'``. The default is None.
+        normalization : str, optional
+            Normalization convention, either ``'n3d'``, ``'maxN'`` or ``'sn3d'``. The default is ``'n3d'``.
+            (maxN is only supported up to 3rd order)
+        weights : array-like, optional
+            Sampling weights for the quadrature transform. Required if `quadrature` is chosen.
+        """
+        # TODO: Support for pyfar Coordinates object
+        self.sampling_sphere = spharpy.SamplingSphere.from_pyfar(coords)
+        self.n_max = self.sampling_sphere.n_max
+        self.weights = self.sampling_sphere.weights
+        self.coords = coords  # coordinates : :doc:`pf.Coordinates <pyfar:classes/pyfar.coordinates>`
+        self.basis_type = basis_type
+        self.inv_transform_type = inv_transform_type
+        self.channel_convention = channel_convention  # channel ordering convention
+        self.normalization = normalization  # gain normalization convention
+        self.basis = None
+        self.basis_gradient_theta, self.basis_gradient_phi = (None, None)
+        self._basis_inv = None
+        self._basis_inv_gradient_theta, self._basis_inv_gradient_phi = (None, None)
+    # Properties
+    @property
+    def n_max(self):
+        return self._n_max
+
+    @n_max.setter
+    def n_max(self, value):
+        if value < 0:
+            raise ValueError("n_max must be a positive integer")
+        self._n_max = value
+
+    @property
+    def weights(self):
+        """Sampling weights for numeric integration."""
+        return self._weights
+
+    @weights.setter
+    def weights(self, weights):
+        if weights is None:
+            self._weights = None
+            return
+        if len(weights) != len(self.coords.cartesian):
+            raise ValueError("The number of weights has to be equal to \
+                    the number of sampling points.")
+
```

Comment:
This would be a bit more complicated if we were to allow multidimensional coordinates. In any case, I think we should use `coords.cshape` or `coords.csize` to check this. and also explicitly check the shape of weights.